### PR TITLE
Route experiments

### PR DESCRIPTION
**Pull request by atruskie**
Redesigned routing.

It's a generally better design that is more imperative and requires less searching when the app boots.

It also adds a so-called `StrongRoute` type that can output a single route as a string (`"/security/login"`) for as an array of commands via the `format` method (`["secruity", "login"]`) and will insert route parameters into the right spots.

This patch also enables the Ivy renderer and was checked to be working with the angular compiler aot option.

It, does however, require an update to ngx-formly: https://github.com/ngx-formly/ngx-formly/issues/1693

### CODE REVIEW

File: src/app/component/shared/menu/menu.component.html [hunk 1]
```diff
@@ -24,15 +24,15 @@ <h6 class="nav-link disabled m-0">
...
+        <ng-container *ngIf="isInternalLink(link)">
           <app-menu-internal-link
             [id]="menuType + '-tooltip-' + i"
             [link]="link"
             [placement]="placement"
           >
           </app-menu-internal-link>
         </ng-container>
-        <ng-template #elseExternalLink>
+        <ng-template *ngIf="isExternalLink(link)">
```
Comment (Allcharles) [inline src/app/component/shared/menu/menu.component.html]: This should be an `ng-container`. Reason why is here: https://stackoverflow.com/questions/40529537/ng2-difference-between-ng-container-and-ng-template-tags

File: angular.json [hunk 2]
```diff
@@ -22,8 +22,8 @@
             "main": "src/main.ts",
             "polyfills": "src/polyfills.ts",
             "tsConfig": "tsconfig.app.json",
-            "aot": false,
-            "assets": ["src/assets"],
+            "aot": true,
```
Comment (Allcharles) [inline angular.json:25]: I can see from the tsconfig that you are using ivy which requires AOT. However, we should have a separate build process for developing if possible as AOT disables the debugging tools such as Augury.

File: src/app/component/error/errorModule.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      {
+        path: "**",
+        component: PageNotFoundComponent
+      }
+    ])
+  ],
+  declarations: [PageNotFoundComponent],
+  exports: [PageNotFoundComponent, RouterModule]
+})
+export class ErrorModule {}
```
Comment (Allcharles) [inline src/app/component/error/errorModule.ts]: I am assuming you wish to put more components into the error module?
  - Page not found
    - For this, are we going to attempt to use it in a similar manner to ecosounds or create a different screen?
  - Unauthorised access
  - Resource has moved?

File: src/app/interfaces/Routing.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                } else {
+                    throw new Error(`Parameter named ${x.name} was not supplied a value and a default value was not given`);
+                }
+            } else {
+                return x.name;
+            }
+        };
+
+        return this.full.map(prepareParam);
+    }
```
Comment (Allcharles) [inline src/app/interfaces/Routing.ts]: `format` function is never used

File: src/app/interfaces/Routing.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            // generate it.
+            callback(current.pageComponent, current.config);
+            const thisRoute = current.routeConfig;
+            const childRoutes = current.children.map(recursiveAdd);
+            // TODO: the order of these arrays may well need to be changed...
+            // depends on how nested routes work
+            // we want "" to match current
+            // but e.g. "sites" to go down a level
+            // and not for both to match at same time.
+            thisRoute.children = [...(thisRoute.children || []), ...childRoutes];
```
Comment (Allcharles) [inline src/app/interfaces/Routing.ts]: Would this not just be a simple sort function? eg. `If string.length === 0 return higher priority, else sort by alphabet`?

**Review by Allcharles — APPROVED** [review]
Changing the names of default files (eg. app-routing.module.ts) may cause some confusion if a new angular developer works on this project.
There are a few bugs stopping the project from building but I've already worked out most of them (mostly just broken imports).

File: angular.json (hunk 2 above)
Comment (atruskie) [reply angular.json:25]: Yeah, I saw that augury didn't work with AOT :-( 

The [Ivy guide](https://angular.io/guide/ivy) suggests always using AOT:

>AOT compilation with Ivy is faster and should be used by default

I'll look into adding another build.

File: src/app/component/error/errorModule.ts (hunk 3 above)
Comment (atruskie) [reply src/app/component/error/errorModule.ts]: Those are good ideas.

Mainly I wanted 404s to render a page on the faulty URL and not redirect to `/`. I just tried to follow the convention we had set up for other components/modules.

File: src/app/interfaces/Routing.ts (hunk 4 above)
Comment (atruskie) [reply src/app/interfaces/Routing.ts]: Never used yet. It's designed to be used in templates. It is the most useful for parameterized routes, of which we did not have any yet in this branch. I'll update some of the templates...

File: src/app/interfaces/Routing.ts (hunk 5 above)
Comment (atruskie) [reply src/app/interfaces/Routing.ts]: Well, there's no need to sort when you can place the items in the order you want.

What I'm unsure of is the order they should be. I suspect we'll find out quickly as soon as we add a nested route.

**Comment by atruskie** [comment]
> Changing the names of default files (eg. app-routing.module.ts) may cause some confusion if a new angular developer works on this project.

Ah, I didn't realise this was a default name.

Standard naming of JS files should be pascalCase. I get angular components being kebab-case because that is the same name as their html tag. However, since app-routing was not a component...

**Comment by Allcharles** [comment]
Do you know if this will address the issue #5? The biggest issue with the old system was that a component could not reference its parent or itself in the action menu (and partially the secondary menu which would break the ability for breadcrumbs).

**Comment by atruskie** [comment]
> Do you know if this will address the issue #5? The biggest issue with the old system was that a component could not reference its parent or itself in the action menu (and partially the secondary menu which would break the ability for breadcrumbs).

I intended it to solve #5, but didn't check. Will follow up. 

In #5 can you add an example configuration that did not work?

**Review by Allcharles — APPROVED** [review]

**Comment by Allcharles** [comment]
Currently receiving the following error when I try to npm start:
`ERROR in Cannot read property 'loadChildren' of undefined`

**Comment by atruskie** [comment]
yup, still working on it. None of the solutions I've found so far are adequate. This problem does not happen with the Ivy renderer...

Don't merge yet

**Pull request merged**